src/Network/Wai/SAML2/Assertion.hs

```diff
@@ -230,3 +232,11 @@ instance FromXML Assertion where
         }
 
 --------------------------------------------------------------------------------
+
+-- | Represents a SAML2 assertion.
+data AssertionWithState = AssertionWithState {
+    -- | Assertion
+    assertion :: !Assertion,
+    -- | Relay state, as provided in the POST req
+    relayState :: !(Maybe ByteString)
+} deriving (Eq, Show)
```

Comment:
I think this would be better off in the `Network.Wai.SAML2` module, than this module. 
```suggestion
-- | Represents the result of validating a SAML2 response.
data Result = Result {
    -- | The assertion obtained from the response that has been validated.
    assertion :: !Assertion,
    -- | An optional relay state, as provided in the POST request.
    relayState :: !(Maybe BS.ByteString)
} deriving (Eq, Show)
```